network-protection/network-protection-internal/src/main/java/com/duckduckgo/networkprotection/internal/feature/snooze/VpnCallStateReceiver.kt

```diff
@@ -0,0 +1,205 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.networkprotection.internal.feature.snooze
+
+import android.annotation.SuppressLint
+import android.content.*
+import android.telephony.PhoneStateListener
+import android.telephony.TelephonyManager
+import androidx.core.content.edit
+import androidx.lifecycle.LifecycleOwner
+import com.duckduckgo.anvil.annotations.InjectWith
+import com.duckduckgo.app.di.AppCoroutineScope
+import com.duckduckgo.app.di.ProcessName
+import com.duckduckgo.app.global.DispatcherProvider
+import com.duckduckgo.app.lifecycle.MainProcessLifecycleObserver
+import com.duckduckgo.di.scopes.AppScope
+import com.duckduckgo.di.scopes.ReceiverScope
+import com.duckduckgo.mobile.android.vpn.Vpn
+import com.duckduckgo.mobile.android.vpn.prefs.VpnSharedPreferencesProvider
+import com.squareup.anvil.annotations.ContributesBinding
+import com.squareup.anvil.annotations.ContributesMultibinding
+import javax.inject.Inject
+import kotlin.properties.Delegates
+import kotlinx.coroutines.*
+import logcat.LogPriority
+import logcat.asLog
+import logcat.logcat
+
+interface VpnDisableOnCall {
+    fun enable()
+    fun disable()
+
+    suspend fun isEnabled(): Boolean
+}
+
+@InjectWith(ReceiverScope::class)
+@ContributesMultibinding(
+    scope = AppScope::class,
+    boundType = MainProcessLifecycleObserver::class,
+)
+@ContributesBinding(
+    scope = AppScope::class,
+    boundType = VpnDisableOnCall::class,
+)
+class VpnCallStateReceiver @Inject constructor(
+    private val context: Context,
+    private val vpn: Vpn,
+    private val dispatcherProvider: DispatcherProvider,
+    @AppCoroutineScope private val appCoroutineScope: CoroutineScope,
+    private val sharedPreferencesProvider: VpnSharedPreferencesProvider,
+    @ProcessName private val processName: String,
+) : BroadcastReceiver(), MainProcessLifecycleObserver, VpnDisableOnCall {
+
+    private val preferences: SharedPreferences by lazy {
+        sharedPreferencesProvider.getSharedPreferences(PREFS_FILENAME, multiprocess = true, migrate = false)
+    }
+
+    private val telephonyManager by lazy {
+        context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager?
+    }
+    private val _listener: PhoneStateListener =
+        object : PhoneStateListener() {
+            @Deprecated("Deprecated in Java")
+            override fun onCallStateChanged(state: Int, phoneNumber: String?) {
+                appCoroutineScope.launch(dispatcherProvider.io()) {
+                    logcat { "Call state: $state" }
```

Comment:
Note that regardless the state of the VPN, this is being called as long as the internal setting is enabled. I think we should only call this if the vpnn has been enabled prior.

I see that when VPN is disabled every after a call:

- VPN log: onCreate, creating the APP_TP_VPN network stack
- VPN log: VPN network APP_TP_VPN created
- VPN log: onStartCommand: ACTION_START_VPN
- VPN log: onDestroy

Reply:
Yeah, the difficulty I found doing that eg. in vpn callbacks is that the receiver needs to outlive the VPN service (needs to re-enable the VPN once call is done).
Lmk if you have ideas, open to them